### PR TITLE
NOISSUE - Make application/json content-type valid in http-adapter

### PR DESCRIPTION
**Pull request by manuio**
Signed-off-by: Manuel Imperiale <manuel.imperiale@gmail.com>

### What does this do?
Make application/json content-type valid in http-adapter

### List any changes that modify/break current functionality
N/A
### Have you included tests for your changes?
yes
### Did you document any new/modified functionality?
no

### CODE REVIEW

**Review by mteodor — APPROVED** [review]

**Review by mteodor — DISMISSED** [review]

**Comment by drasko** [comment]
Why is this PR needed? Because of generic JSON writer?

@dusanb94 please comment on this PR, this looks strange to me. It is impractical to enter one by one content type, IMHO Mainflux is for now enabled only with SenML protocol knowledge.

**Comment by manuio** [comment]
> Why is this PR needed? Because of generic JSON writer?
> 
> @dusanb94 please comment on this PR, this looks strange to me. It is impractical to enter one by one content type, IMHO Mainflux is for now enabled only with SenML protocol knowledge.

@drasko Yes it'sbecause of the generic JSON writer. Currently the http-adapter only accept `application/senml+json`If you want to send a payload with pure JSON you have to use the same content-type in the header and it's not really consistent. This PR only make both types valid. BTW, I should maybe add `application/senml+cbor`. Or should I remove content-type check?

**Review by dborovcanin — DISMISSED** [review]

**Comment by dborovcanin** [comment]
> > Why is this PR needed? Because of generic JSON writer?
> > @dusanb94 please comment on this PR, this looks strange to me. It is impractical to enter one by one content type, IMHO Mainflux is for now enabled only with SenML protocol knowledge.
> 
> @drasko Yes it'sbecause of the generic JSON writer. Currently the http-adapter only accept `application/senml+json`If you want to send a payload with pure JSON you have to use the same content-type in the header and it's not really consistent. This PR only make both types valid. BTW, I should maybe add `application/senml+cbor`. Or should I remove content-type check?

TBH, adding CBOR CT also makes sense, since we support, well, all three.

File: http/api/endpoint_test.go [hunk 1]
```diff
@@ -83,36 +87,50 @@ func TestPublish(t *testing.T) {
...
+			msg:         msgJSON,
+			contentType: contentTypeJSON,
 			key:         thingKey,
 			status:      http.StatusAccepted,
 		},
 		"publish message with empty key": {
 			chanID:      chanID,
 			msg:         msg,
-			contentType: contentType,
+			contentType: contentTypeSenML,
```
Comment (drasko) [inline http/api/endpoint_test.go]: Why only SenML, what about the other 3 CTs we support?

File: http/api/transport.go [hunk 2]
```diff
@@ -176,7 +180,7 @@ func encodeError(_ context.Context, err error, w http.ResponseWriter) {
 	}
 
 	if errorVal, ok := err.(errors.Error); ok {
-		w.Header().Set("Content-Type", contentType)
+		w.Header().Set("Content-Type", contentTypeJSON)
```
Comment (drasko) [inline http/api/transport.go]: Why only JSON, what about other 3 CTs we support?

File: http/api/endpoint_test.go [hunk 3]
```diff
@@ -63,10 +63,14 @@ func (tr testRequest) make() (*http.Response, error) {
 
 func TestPublish(t *testing.T) {
 	chanID := "1"
-	contentType := "application/senml+json"
+	contentTypeSenML := "application/senml+json"
+	contentTypeCBOR := "application/senml+cbor"
```
Comment (drasko) [inline http/api/endpoint_test.go]: Looks like naming is not consistent. If we call upper SenML (should be maybe `SenMLJSON`), this one is also SenML (just that it is CBOR encoded). So in the upper we use semantic representation (SenML) and in the second we use encoding/serialization (although semantic structures are still the same).

**Review by drasko — CHANGES_REQUESTED** [review]

File: http/api/endpoint_test.go (hunk 1 above)
Comment (manuio) [reply http/api/endpoint_test.go]: There are already tests for CTs validation and it's pointless to test 3 CTs for `empty key`

File: http/api/transport.go (hunk 2 above)
Comment (manuio) [reply http/api/transport.go]: Because it's the CT of the error response

File: http/api/transport.go [hunk 4]
```diff
@@ -96,7 +98,9 @@ func parseSubtopic(subtopic string) (string, error) {
 }
 
 func decodeRequest(ctx context.Context, r *http.Request) (interface{}, error) {
-	if !strings.Contains(r.Header.Get("Content-Type"), contentType) {
+	if !strings.Contains(r.Header.Get("Content-Type"), contentTypeSenMLJSON) &&
+		!strings.Contains(r.Header.Get("Content-Type"), contentTypeSenMLCBOR) &&
+		!strings.Contains(r.Header.Get("Content-Type"), contentTypeJSON) {
```
Comment (dborovcanin) [inline http/api/transport.go]: Extract to a variable `ct := r.Header.Get("Content-Type")`. Also, why don't we compare with `==` rather than using `strings.Contains`?

File: http/api/transport.go [hunk 5]
```diff
@@ -29,8 +29,10 @@ import (
 )
 
 const (
-	protocol    = "http"
-	contentType = "application/senml+json"
+	protocol         = "http"
+	contentTypeSenMLJSON = "application/senml+json"
+	contentTypeSenMLCBOR  = "application/senml+cbor"
+	contentTypeJSON  = "application/json"
```
Comment (dborovcanin) [inline http/api/transport.go]: Consider shortening to `ctSenmlJSON`, `ctSenmlCBOR`, `ctJSON` (I know it's `SenML` and not `Senml`, but this looks more natural; and since it's a local variable used only once, `ct` instead of `contentType` sounds acceptable).

File: pkg/sdk/go/message.go [hunk 6]
```diff
@@ -21,7 +21,7 @@ func (sdk mfSDK) SendMessage(chanName, msg, key string) error {
 		subtopicPart = fmt.Sprintf("/%s", strings.Replace(chanNameParts[1], ".", "/", -1))
 	}
 
-	url := fmt.Sprintf("%s/channels/%s/messages%s", sdk.httpAdapterURL, chanID, subtopicPart)
+	url := fmt.Sprintf("%s/channels/%s/messages/json%s", sdk.httpAdapterURL, chanID, subtopicPart)
```
Comment (dborovcanin) [inline pkg/sdk/go/message.go]: Why did you add this subtopic?

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Please remove the binary file from `cmd/cli`.

**Review by dborovcanin — DISMISSED** [review]

**Comment by dborovcanin** [comment]
@manuio Please remove the binary from `/cmd/cli`.

**Review by drasko — APPROVED** [review]
LGTM

**Review by dborovcanin — APPROVED** [review]
